core/coreapi/pin.go

```diff
@@ -194,57 +195,119 @@ func (p *pinInfo) Type() string {
 }
 
 func (api *PinAPI) pinLsAll(typeStr string, ctx context.Context) ([]coreiface.Pin, error) {
+	pinCh, errCh := PinLsAll(ctx, typeStr, api.pinning, api.dag)
+
+	var pins []coreiface.Pin
+loop:
+	for {
+		select {
+		case p, ok := <-pinCh:
+			if !ok {
+				break loop
+			}
+			pins = append(pins, p)
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+	}
+	err := <-errCh
+	if err != nil {
+		return nil, err
+	}
+
+	return pins, nil
+}
 
-	keys := make(map[cid.Cid]*pinInfo)
+// PinLsAll is an internal function for returning a list of pins
+func PinLsAll(ctx context.Context, typeStr string, pin pin.Pinner, dag ipld.DAGService) (chan coreiface.Pin, chan error) {
+	ch := make(chan coreiface.Pin, 32)
+	errCh := make(chan error, 1)
 
-	AddToResultKeys := func(keyList []cid.Cid, typeStr string) {
+	keys := cid.NewSet()
+	AddToResultKeys := func(keyList []cid.Cid, typeStr string) error {
 		for _, c := range keyList {
-			keys[c] = &pinInfo{
-				pinType: typeStr,
-				path:    path.IpldPath(c),
+			if keys.Visit(c) {
+				select {
+				case ch <- &pinInfo{
+					pinType: typeStr,
+					path:    path.IpldPath(c),
+				}:
+				case <-ctx.Done():
+					return ctx.Err()
+				}
 			}
 		}
+		return nil
 	}
 
-	if typeStr == "direct" || typeStr == "all" {
-		dkeys, err := api.pinning.DirectKeys(ctx)
-		if err != nil {
-			return nil, err
-		}
-		AddToResultKeys(dkeys, "direct")
-	}
-	if typeStr == "indirect" || typeStr == "all" {
-		set := cid.NewSet()
-		rkeys, err := api.pinning.RecursiveKeys(ctx)
-		if err != nil {
-			return nil, err
-		}
-		for _, k := range rkeys {
-			err := merkledag.Walk(
-				ctx, merkledag.GetLinksWithDAG(api.dag), k,
-				set.Visit,
-				merkledag.SkipRoot(), merkledag.Concurrent(),
-			)
+	go func() {
+		defer close(ch)
+		defer close(errCh)
+		if typeStr == "direct" || typeStr == "all" {
+			dkeys, err := pin.DirectKeys(ctx)
 			if err != nil {
-				return nil, err
+				errCh <- err
+				return
+			}
+			if err := AddToResultKeys(dkeys, "direct"); err != nil {
+				errCh <- err
+				return
 			}
 		}
-		AddToResultKeys(set.Keys(), "indirect")
-	}
-	if typeStr == "recursive" || typeStr == "all" {
-		rkeys, err := api.pinning.RecursiveKeys(ctx)
-		if err != nil {
-			return nil, err
+		if typeStr == "recursive" || typeStr == "all" {
+			rkeys, err := pin.RecursiveKeys(ctx)
+			if err != nil {
+				errCh <- err
+				return
+			}
+			if err := AddToResultKeys(rkeys, "recursive"); err != nil {
+				errCh <- err
+				return
+			}
 		}
-		AddToResultKeys(rkeys, "recursive")
-	}
+		if typeStr == "indirect" || typeStr == "all" {
+			rkeys, err := pin.RecursiveKeys(ctx)
+			if err != nil {
+				errCh <- err
+				return
+			}
 
-	out := make([]coreiface.Pin, 0, len(keys))
-	for _, v := range keys {
-		out = append(out, v)
-	}
+			if typeStr != "all" {
```

Comment:
This needs a comment explaining _why_ we're doing this.

Reply:
I also inverted the test to `if typeStr == "indirect"` to make it clear that we're skipping this because we only care about indirect pins.